c2rust-analyze/src/labeled_ty.rs

```diff
@@ -224,30 +224,28 @@ impl<'tcx, L: Copy> LabeledTyCtxt<'tcx, L> {
     /// Panics if `lty` contains a reference to a type parameter that is past the end of `substs`
     /// (usually this means the caller is providing the wrong list of type arguments as `substs`).
     ///
-    /// TODO: This produces a `LabeledTy` with the right structure, but doesn't actually do
-    /// substitution on the underlying `Ty`s!  This means if you substitute `u32` for `T`, you can
-    /// end up with a `LabeledTy` whose `ty` is `S<T>`, but whose args are `[u32]`.  By some
-    /// miracle, this hasn't broken anything yet, but we may need to fix it eventually.
-    #[allow(dead_code)]
-    pub fn subst(
-        &self,
-        lty: LabeledTy<'tcx, L>,
-        substs: &[LabeledTy<'tcx, L>],
-    ) -> LabeledTy<'tcx, L> {
+    /// This produces a [`LabeledTy`] with the right structure and also substitutes the
+    /// underlying [`Ty`]s
+    pub fn subst(&self, lty: LabeledTy<'tcx, L>, substs: SubstsRef<'tcx>) -> LabeledTy<'tcx, L> {
+        let substituted_ty = EarlyBinder(lty.ty).subst(self.tcx, substs);
+
         if let TyKind::Param(ref ty) = lty.ty.kind() {
-            if let Some(p) = substs.get(ty.index as usize) {
-                return p;
+            if substs.get(ty.index as usize).is_some() {
+                return self.mk(substituted_ty, &[], lty.label);
```

Comment:
This case is incorrect - it produces an `LTy` with empty `args` even if `substituted_ty` ought to have some args.  I think this should return `p` as before, but also assert that `p.ty` matches `substituted_ty`.

Reply:
p is no longer and LTy, though, because the type of substs has changed.

Reply:
Ah, I missed that change.  The argument here should still be a list of `LTy`s, and it should build the `SubstsRef` internally.  The point of this function is to do things like substitute `*mut /*ptr1*/ u32` into `Vec<T>` to get `Vec<*mut /*ptr1*/ u32>`, so the inputs still need to be labeled.

Reply:
addressed in a7152bc

Reply:
No longer used in d8876a7a